Mainboard_ESP32/components/lora/lora.c

```diff
@@ -0,0 +1,295 @@
+
+  uint64_t frf = ((uint64_t)frequency << 19) / 32000000;
+
+  lora_write_reg(lora, REG_FRF_MSB, (uint8_t)(frf >> 16));
+  lora_write_reg(lora, REG_FRF_MID, (uint8_t)(frf >> 8));
+  lora_write_reg(lora, REG_FRF_LSB, (uint8_t)(frf >> 0));
+}
+
+void lora_set_spreading_factor(lora_struct_t *lora, int16_t sf) {
+  if (sf < 6)
+    sf = 6;
+  else if (sf > 12)
+    sf = 12;
+
+  if (sf == 6) {
+    lora_write_reg(lora, REG_DETECTION_OPTIMIZE, 0xc5);
+    lora_write_reg(lora, REG_DETECTION_THRESHOLD, 0x0c);
+  } else {
+    lora_write_reg(lora, REG_DETECTION_OPTIMIZE, 0xc3);
+    lora_write_reg(lora, REG_DETECTION_THRESHOLD, 0x0a);
+  }
+
+  lora_write_reg(
+      lora, REG_MODEM_CONFIG_2,
+      (lora_read_reg(lora, REG_MODEM_CONFIG_2) & 0x0f) | ((sf << 4) & 0xf0));
+}
+
+void lora_set_bandwidth(lora_struct_t *lora, int32_t sbw) {
+  int16_t bw;
+
+  if (sbw <= 7.8E3)
+    bw = 0;
+  else if (sbw <= 10.4E3)
+    bw = 1;
+  else if (sbw <= 15.6E3)
+    bw = 2;
+  else if (sbw <= 20.8E3)
+    bw = 3;
+  else if (sbw <= 31.25E3)
+    bw = 4;
+  else if (sbw <= 41.7E3)
+    bw = 5;
+  else if (sbw <= 62.5E3)
+    bw = 6;
+  else if (sbw <= 125E3)
+    bw = 7;
+  else if (sbw <= 250E3)
+    bw = 8;
+  else
+    bw = 9;
+
+  lora_write_reg(lora, REG_MODEM_CONFIG_1,
+                 (lora_read_reg(lora, REG_MODEM_CONFIG_1) & 0x0f) | (bw << 4));
+}
+
+void lora_set_coding_rate(lora_struct_t *lora, int16_t denominator) {
+  if (denominator < 5)
+    denominator = 5;
+  else if (denominator > 8)
+    denominator = 8;
+
+  int16_t cr = denominator - 4;
+  lora_write_reg(lora, REG_MODEM_CONFIG_1,
+                 (lora_read_reg(lora, REG_MODEM_CONFIG_1) & 0xf1) | (cr << 1));
+}
+
+void lora_set_preamble_length(lora_struct_t *lora, int32_t length) {
+  lora_write_reg(lora, REG_PREAMBLE_MSB, (uint8_t)(length >> 8));
+  lora_write_reg(lora, REG_PREAMBLE_LSB, (uint8_t)(length >> 0));
+}
+
+void lora_set_sync_word(lora_struct_t *lora, int16_t sw) {
+  lora_write_reg(lora, REG_SYNC_WORD, sw);
+}
+
+void lora_enable_crc(lora_struct_t *lora) {
+  lora_write_reg(lora, REG_MODEM_CONFIG_2,
+                 lora_read_reg(lora, REG_MODEM_CONFIG_2) | 0x04);
+}
+
+void lora_disable_crc(lora_struct_t *lora) {
+  lora_write_reg(lora, REG_MODEM_CONFIG_2,
+                 lora_read_reg(lora, REG_MODEM_CONFIG_2) & 0xfb);
+}
+
+// TODO(GLIBUS): Add a method with no delay, also add a breakout from while
+// after some time
+void lora_send_packet(lora_struct_t *lora, uint8_t *buf, int16_t size) {
+  /*
+   * Transfer data to radio.
+   */
+  lora_idle(lora);
+  lora_write_reg(lora, REG_FIFO_ADDR_PTR, 0);
+
+  for (int16_t i = 0; i < size; i++) {
+    lora_write_reg(lora, REG_FIFO, *buf++);
+  }
+  lora_write_reg(lora, REG_PAYLOAD_LENGTH, size);
+
+  /*
+   * Start transmission and wait for conclusion.
+   */
+  lora_write_reg(lora, REG_OP_MODE, MODE_LONG_RANGE_MODE | MODE_TX);
+
+  while ((lora_read_reg(lora, REG_IRQ_FLAGS) & IRQ_TX_DONE_MASK) == 0x00) {
```

Comment:
ja będę chciał tx done na przerwaniu odbierać

Reply:
okej to dodam funkcje na tx done

Reply:
i bardziej funkcje na ktore sie to wysylanie "rozklada"